source/http2_stream_manager.c

```diff
@@ -0,0 +1,987 @@
+            /* Release the acquired connection */
+            re_error |= aws_http_connection_manager_release_connection(stream_manager->connection_manager, connection);
+        } else {
+            struct aws_h2_sm_connection *sm_connection = s_sm_connection_new(stream_manager, connection);
+            bool added = false;
+            re_error |=
+                aws_random_access_set_add(&stream_manager->synced_data.ideal_sm_connection_set, sm_connection, &added);
+            re_error |= !added;
+            ++stream_manager->synced_data.holding_connections_count;
+        }
+        s_aws_http2_stream_manager_build_transaction_synced(&work);
+        s_unlock_synced_data(stream_manager);
+    } /* END CRITICAL SECTION */
+
+    AWS_ASSERT(!re_error && "connection acquired callback fails with programming errors");
+    s_aws_http2_stream_manager_execute_transaction(&work);
+}
+
+static int s_on_incoming_headers(
+    struct aws_http_stream *stream,
+    enum aws_http_header_block header_block,
+    const struct aws_http_header *header_array,
+    size_t num_headers,
+    void *user_data) {
+    struct aws_h2_sm_pending_stream_acquisition *pending_acquisition = user_data;
+    if (pending_acquisition->options.on_response_headers) {
+        return pending_acquisition->options.on_response_headers(
+            stream, header_block, header_array, num_headers, pending_acquisition->options.user_data);
+    }
+    return AWS_OP_SUCCESS;
+}
+
+static int s_on_incoming_header_block_done(
+    struct aws_http_stream *stream,
+    enum aws_http_header_block header_block,
+    void *user_data) {
+    struct aws_h2_sm_pending_stream_acquisition *pending_acquisition = user_data;
+    if (pending_acquisition->options.on_response_header_block_done) {
+        return pending_acquisition->options.on_response_header_block_done(
+            stream, header_block, pending_acquisition->options.user_data);
+    }
+    return AWS_OP_SUCCESS;
+}
+
+static int s_on_incoming_body(struct aws_http_stream *stream, const struct aws_byte_cursor *data, void *user_data) {
+    struct aws_h2_sm_pending_stream_acquisition *pending_acquisition = user_data;
+    if (pending_acquisition->options.on_response_body) {
+        return pending_acquisition->options.on_response_body(stream, data, pending_acquisition->options.user_data);
+    }
+    return AWS_OP_SUCCESS;
+}
+
+/* Helper invoked when underlying connections is still aviable and the num stream assigned has been updated */
+static int s_update_sm_connection_set_on_stream_finishes_synced(
+    struct aws_h2_sm_connection *sm_connection,
+    struct aws_http2_stream_manager *stream_manager) {
+    int re_error = 0;
+    size_t cur_num = sm_connection->num_streams_assigned;
+    size_t ideal_num = stream_manager->ideal_concurrent_streams_per_connection;
+    size_t max_num = sm_connection->max_concurrent_streams;
+    if (sm_connection->sim_full && cur_num < ideal_num && ideal_num < max_num) {
```

Comment:
mayyyyybe: remove `&& ideal_num < max_num`?

Reply:
also: add TODO somewhere about how MAX_CONCURRENT_STREAMS changing unexpectedly would really mess us up

logic would need to be like:
- figure out where I am
- figure out where I should be
- if they're different, remove from where I am, put where should be